gcpvault.go

```diff
@@ -315,3 +453,21 @@ func getEmailFromCredentials(creds *google.Credentials) (string, error) {
 
 	return data["client_email"], nil
 }
+
+func isExpired(token *Token, cfg Config) bool {
+	if token == nil {
+		return true
+	}
+
+	refreshTime := time.Now().Add(time.Minute * time.Duration(cfg.CachedTokenRefreshThreshold))
+	//seed random generator
+	rand.Seed(time.Now().UnixNano())
+	//subtract random number of seconds from the expiration to avoid many simultaneous refresh events
+	refreshTime = refreshTime.Add(time.Second * (-1 * time.Duration(rand.Intn(60))))
```

Comment:
I'm not sure I understand the point of this part?

Since the refresh time comparison is only a couple seconds - the threshold time, when new instances come up close to the expiration, some will use the still valid cached token (great) while others will ask for a new token (not really necessary when the existing one is still valid) which would still cause multiple refresh events(?). The random jitter here doesn't change _when_ the API calls to vault are made, it just changes _how many_ got lucky with the expire timestamp.

And in the case that the cached token is actually expired or doesn't exist, all clients would stampede the vault client.

I think if you want to add jitter, it needs to be at the program initialization so that there's jitter when checking the cache, which allows _some_ instance of an app to have updated the cache with a vault token that is unexpired/has a long expiration time.

Another way I can think of implementing a token refresh routine is like this:
- on startup it checks the cache - if exists && still unexpired => use it, if it doesn't exist || expired => call vault
- after app is running, run a goroutine in the background that performs token refreshes with longer thresholds (aka 1 hour) and a bigger jitter

This should also eliminate the problem when new instances scale up because there is still a valid token in cache as long as 1 instance was long running. the only time multiple instances would call vault is if 0 instances existed prior and there is no jitter in the initial call to the cache or vault.

Let me know if that makes sense.

Reply:
FYI. @tonglil have discussed this and moved our conversation to the RFC.
https://docs.google.com/document/d/16vJFwr-rWasf_XvkGwYdK6D_BCbt0YcWpKhnzxWFEcc/edit#